File: classes/选择&冒泡&插入排序/sort.py
# 选择排序 i ~ n - 1 找到最小值，与i位置交换
def select_sort(arr):
    n = len(arr)
    for i in range(n):
        min_index = i
        for j in range(i + 1, n):
            min_index = j if arr[j] < arr[min_index] else min_index
        arr[i], arr[min_index] = arr[min_index], arr[i]

# 冒泡排序, end = n - 1, 0 ~ end 找寻最大的，最大的移到end, end--
def bubble_sort(arr):
    end = len(arr) - 1
    for i in range(end):
        for j in range(end - i):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]


# 插入排序 0 ～ i 范围有序， i + 1 插入到 0 ~ i 范围中排序后的位置
def insert_sort(arr):
    for i in range(1, len(arr)):
        for j in range(i - 1, -1, -1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]

File: classes/选择/冒泡/插入排序/sort.js
function swap(arr, i , j) {
    let t = arr[i];
    arr[i] = arr[j];
    arr[j] = t;
}

// 选择排序 i ~ n - 1 找到最小值，与i位置交换
function selectSort(arr) {
    for (let i = 0; i < arr.length; i ++) {
        let minIndex = i;
        for (let j = i + i; j < arr.length; j ++) {
            minIndex = arr[j] < arr[minIndex] ? j : minIndex;
        }
        swap(arr, i, minIndex);
    }
}

// 冒泡排序, end = n - 1, 0 ~ end 找寻最大的，最大的移到end, end--
function bubbleSort(arr) {
    let end = arr.length - 1;
    for (let i = 0; i < end; i ++) {
        for (let j = 0; j < end - i; j ++) {
            if (arr[j] > arr[j + 1]) {
                swap(arr, j, j + 1);
            }
        }
    }
}

// 插入排序 0 ～ i 范围有序， i + 1 插入到 0 ~ i 范围中排序后的位置
function insertSort(arr) {
    for (let i = 1; i < arr.length; i ++) {
        for (let j = i - 1; j >= 0 && arr[j] > arr[j + 1]; j --) {
            swap(arr, j, j + 1);
        }
    }
}